pkg/codegen/go/gen_program_expressions.go

```diff
@@ -814,9 +858,14 @@ func (g *generator) genApply(w io.Writer, expr *model.FunctionCallExpression) {
 	isInput := false
 	retType := g.argumentTypeName(nil, then.Signature.ReturnType, isInput)
 	// TODO account for outputs in other namespaces like aws
-	typeAssertion := fmt.Sprintf(".(%sOutput)", retType)
-	if !strings.HasPrefix(retType, "pulumi.") {
-		typeAssertion = fmt.Sprintf(".(pulumi.%sOutput)", Title(retType))
+	var typeAssertion string
+	if retType == "[]string" {
+		typeAssertion = ".(pulumi.StringArrayOutput)"
```

Comment:
This is far from complete here, I'm adding just enough to pass the tests. Any thoughts on how to factor this out? We need to backlog this.

Reply:
Not offhand. Definitely need to backlog this. In general we need to go take another hard look at the Go program generator and its approach to naming.